File: MacApp/MacWorkspaceSurfaceActual.vb
Imports AppKit
Imports CoreGraphics
Imports Foundation
Imports QuartzCore
Imports RemObjects.Elements.RTL
Imports [Shared].Core

Namespace Application

  Public Class MacWorkspaceSurfaceActual
    Implements IWorkspaceSurfaceActual

    Public Property CommandRequested As WorkspaceCommandRequested
    Public Property DismissRequested As WorkspaceSurfaceRequested
    Public Property TargetDisplayDisconnected As WorkspaceSurfaceRequested

    Private _window As MacWorkspaceWindow
    Private _imageView As MacWorkspaceImageView
    Private _hudLabel As NSTextField
    Private _ownedFrame As MacWorkspaceFrame
    Private _targetDisplayId As String
    Private _screenObserver As NSObject
    Private _windowCloseObserver As NSObject
    Private _hudFadeTimer As NSTimer
    Private _isDismissing As Boolean
    Private _didRequestDisplayDismissal As Boolean

    Public Function presentFrame(frame workspaceFrame As IWorkspaceFrame, onDisplay display As WorkspaceDisplay) As WorkspacePresentationResult
      If Not NSThread.isMainThread Then
        Return failedPresentationWithMessage("工作区界面必须在主线程显示")
      End If

      If _window IsNot Null Then
        Return failedPresentationWithMessage("工作区已经显示")
      End If

      Dim nativeFrame As MacWorkspaceFrame = TryCast(workspaceFrame, MacWorkspaceFrame)
      If nativeFrame Is Null OrElse nativeFrame.IsReleased() Then
        Return failedPresentationWithMessage("无法显示工作区：截图无效")
      End If

      If display Is Null Then
        Return failedPresentationWithMessage("无法显示工作区：显示器信息无效")
      End If

      Dim targetScreen As NSScreen = screenWithIdentifier(display.DisplayId())
      If targetScreen Is Null Then
        Return failedPresentationWithMessage("无法显示工作区：目标显示器不可用")
      End If

      Dim image As NSImage = nativeFrame.imageForPresentation()
      If image Is Null Then
        Return failedPresentationWithMessage("无法显示工作区：截图已经释放")
      End If

      Dim newWindow As MacWorkspaceWindow = Null
      Try
        Dim screenFrame As NSRect = targetScreen.frame
        newWindow = New MacWorkspaceWindow(contentRect: screenFrame,
          styleMask: NSWindowStyleMask.Borderless,
          backing: NSBackingStoreType.Buffered,
          defer: False,
          screen: targetScreen)
        newWindow.backgroundColor = NSColor.blackColor
        newWindow.opaque = True
        newWindow.acceptsMouseMovedEvents = True

        Dim contentFrame As NSRect = NSMakeRect(0.0, 0.0, screenFrame.size.width, screenFrame.size.height)
        Dim contentView As NSView = New NSView(Frame: contentFrame)
        Dim imageView As MacWorkspaceImageView = New MacWorkspaceImageView(Frame: contentFrame)
        imageView.Image = image
        imageView.wantsLayer = True
        contentView.addSubview(imageView)

        Dim hudLabel As NSTextField = NSTextField.labelWithString("100%")
        hudLabel.frame = NSMakeRect(16.0, 16.0, 88.0, 36.0)
        hudLabel.font = NSFont.monospacedDigitSystemFontOfSize(18.0, weight: NSFontWeightSemibold)
        hudLabel.textColor = NSColor.whiteColor
        hudLabel.alignment = NSTextAlignment.Center
        hudLabel.wantsLayer = True
        hudLabel.layer.backgroundColor = NSColor.colorWithWhite(0.0, alpha: 0.72).CGColor
        hudLabel.layer.cornerRadius = 7.0
        hudLabel.hidden = True
        contentView.addSubview(hudLabel)

        newWindow.contentView = contentView
        imageView.CommandRequested = AddressOf workspaceCommandRequested
        imageView.DismissRequested = AddressOf workspaceDismissRequested

        _window = newWindow
        _imageView = imageView
        _hudLabel = hudLabel
        _ownedFrame = nativeFrame
        _targetDisplayId = display.DisplayId()
        _didRequestDisplayDismissal = False
        beginObservingDisplayForWindow(newWindow, displayId: _targetDisplayId)

        NSApplication.sharedApplication().activateIgnoringOtherApps(True)
        newWindow.makeKeyAndOrderFront(Null)
        newWindow.makeFirstResponder(imageView)
        Return WorkspacePresentationResult.succeeded()
      Catch caughtError As Exception
        _isDismissing = True
        stopObservingDisplayChanges()
        If newWindow IsNot Null Then
          newWindow.close()
        End If

        clearPresentationWithoutReleasingFrame()
        _isDismissing = False
        Return failedPresentationWithMessage("无法显示工作区")
      End Try
    End Function

    Public Sub renderTransform(transform As WorkspaceTransform, showHud showHud As Boolean)
      If Not NSThread.isMainThread Then
        NSOperationQueue.mainQueue().addOperationWithBlock(Sub()
          renderTransform(transform, showHud: showHud)
        End Sub)
        Exit Sub
      End If

      If _window Is Null OrElse _imageView Is Null OrElse transform Is Null Then
        Exit Sub
      End If

      _imageView.renderTransform(transform)

      If _hudLabel IsNot Null Then
        _hudLabel.stringValue = NSString.stringWithFormat("%.0f%%", transform.Scale() * 100.0)
        If showHud Then
          showHudTemporarily()
        End If
      End If
    End Sub

    Public Sub dismissPresentation()
      If Not NSThread.isMainThread Then
        NSOperationQueue.mainQueue().addOperationWithBlock(Sub()
          dismissPresentation()
        End Sub)
        Exit Sub
      End If

      _isDismissing = True
      stopObservingDisplayChanges()
      stopHudFadeTimer()

      If _imageView IsNot Null Then
        _imageView.CommandRequested = Null
        _imageView.DismissRequested = Null
      End If

      If _window IsNot Null Then
        _window.orderOut(Null)
        _window.close()
      End If

      releaseOwnedFrame()
      clearPresentationWithoutReleasingFrame()
      _isDismissing = False
    End Sub

    Private Sub workspaceCommandRequested(command As WorkspaceCommand)
      Dim commandListener As WorkspaceCommandRequested = CommandRequested()
      If assigned(commandListener) Then
        commandListener(command)
      End If
    End Sub

    Private Sub showHudTemporarily()
      stopHudFadeTimer()
      _hudLabel.hidden = False
      _hudLabel.alphaValue = 1.0
      _hudFadeTimer = NSTimer.scheduledTimerWithTimeInterval(0.8,
        repeats: False,
        block: Sub(timer)
          _hudFadeTimer = Null
          If _hudLabel IsNot Null Then
            NSAnimationContext.runAnimationGroup(Sub(animation)
              animation.duration = 0.2
              _hudLabel.animator().alphaValue = 0.0
            End Sub,
            completionHandler: Null)
          End If
        End Sub)
    End Sub

    Private Sub stopHudFadeTimer()
      If _hudFadeTimer IsNot Null Then
        _hudFadeTimer.invalidate()
        _hudFadeTimer = Null
      End If
    End Sub

    Private Sub workspaceDismissRequested()
      requestDismissal()
    End Sub

    Private Sub beginObservingDisplayForWindow(window As NSWindow, displayId targetDisplayId As String)
      Dim center As NSNotificationCenter = NSNotificationCenter.defaultCenter()
      _screenObserver = center.addObserverForName(NSApplicationDidChangeScreenParametersNotification,
        [object]: Null,
        queue: NSOperationQueue.mainQueue(),
        usingBlock: Sub(notification)
          checkTargetDisplayAvailability()
        End Sub)
      _windowCloseObserver = center.addObserverForName(NSWindowWillCloseNotification,
        [object]: window,
        queue: NSOperationQueue.mainQueue(),
        usingBlock: Sub(notification)
          If Not _isDismissing Then
            requestDismissal()
          End If
        End Sub)
    End Sub

    Private Sub stopObservingDisplayChanges()
      Dim center As NSNotificationCenter = NSNotificationCenter.defaultCenter()
      If _screenObserver IsNot Null Then
        center.removeObserver(_screenObserver)
        _screenObserver = Null
      End If

      If _windowCloseObserver IsNot Null Then
        center.removeObserver(_windowCloseObserver)
        _windowCloseObserver = Null
      End If
    End Sub

    Private Sub checkTargetDisplayAvailability()
      If _targetDisplayId Is Null OrElse _didRequestDisplayDismissal Then
        Exit Sub
      End If

      If screenWithIdentifier(_targetDisplayId) IsNot Null Then
        Exit Sub
      End If

      _didRequestDisplayDismissal = True
      Dim disconnectListener As WorkspaceSurfaceRequested = TargetDisplayDisconnected()
      If assigned(disconnectListener) Then
        disconnectListener()
      End If
    End Sub

    Private Function screenWithIdentifier(displayId As String) As NSScreen
      If displayId Is Null Then
        Return Null
      End If

      For Each screen As NSScreen In NSScreen.screens
        If identifierForScreen(screen) = displayId Then
          Return screen
        End If
      Next

      Return Null
    End Function

    Private Function identifierForScreen(screen As NSScreen) As String
      Dim displayNumber As NSNumber = TryCast(screen.deviceDescription.objectForKey("NSScreenNumber"), NSNumber)
      If displayNumber Is Null Then
        Return Null
      End If

      Return displayNumber.stringValue
    End Function

    Private Function failedPresentationWithMessage(message As String) As WorkspacePresentationResult
      Return WorkspacePresentationResult.failedWithFailure(New WorkspaceFailure(WorkspaceFailureCode.PresentationFailed,
        message: message))
    End Function

    Private Sub releaseOwnedFrame()
      If _ownedFrame IsNot Null Then
        _ownedFrame.releaseFrame()
        _ownedFrame = Null
      End If
    End Sub

    Private Sub clearPresentationWithoutReleasingFrame()
      _window = Null
      _imageView = Null
      _hudLabel = Null
      _ownedFrame = Null
      _targetDisplayId = Null
      _didRequestDisplayDismissal = False
    End Sub

    Private Sub requestDismissal()
      Dim dismissListener As WorkspaceSurfaceRequested = DismissRequested()
      If assigned(dismissListener) Then
        dismissListener()
      End If
    End Sub

  End Class

End Namespace

File: MacApp/MacWorkspaceSurfaceActual.pas
namespace MacApp;

interface

uses
  AppKit,
  CoreGraphics,
  Foundation,
  QuartzCore,
  RemObjects.Elements.RTL,
  Core;

type
  MacWorkspaceSurfaceActual = public class(IWorkspaceSurfaceActual)
  private
    fCommandRequested: WorkspaceCommandRequested;
    fDismissRequested: WorkspaceSurfaceRequested;
    fTargetDisplayDisconnected: WorkspaceSurfaceRequested;
    fWindow: MacWorkspaceWindow;
    fImageView: MacWorkspaceImageView;
    fHudLabel: NSTextField;
    fOwnedFrame: MacWorkspaceFrame;
    fTargetDisplayId: String;
    fScreenObserver: NSObject;
    fWindowCloseObserver: NSObject;
    fHudFadeTimer: NSTimer;
    fIsDismissing: Boolean;
    fDidRequestDisplayDismissal: Boolean;

    method workspaceCommandRequested(command: WorkspaceCommand);
    method showHudTemporarily;
    method stopHudFadeTimer;
    method workspaceDismissRequested;
    method beginObservingDisplayForWindow(window: NSWindow) displayId(targetDisplayId: String);
    method stopObservingDisplayChanges;
    method checkTargetDisplayAvailability;
    method screenWithIdentifier(displayId: String): NSScreen;
    method identifierForScreen(screen: NSScreen): String;
    method failedPresentationWithMessage(message: String): WorkspacePresentationResult;
    method releaseOwnedFrame;
    method clearPresentationWithoutReleasingFrame;
    method requestDismissal;
  public
    property CommandRequested: WorkspaceCommandRequested read fCommandRequested write fCommandRequested;
    property DismissRequested: WorkspaceSurfaceRequested read fDismissRequested write fDismissRequested;
    property TargetDisplayDisconnected: WorkspaceSurfaceRequested read fTargetDisplayDisconnected write fTargetDisplayDisconnected;

    method presentFrame(workspaceFrame: IWorkspaceFrame) onDisplay(display: WorkspaceDisplay): WorkspacePresentationResult;
    method renderTransform(transform: WorkspaceTransform) showHud(showHud: Boolean);
    method dismissPresentation;
  end;

implementation

method MacWorkspaceSurfaceActual.presentFrame(workspaceFrame: IWorkspaceFrame) onDisplay(display: WorkspaceDisplay): WorkspacePresentationResult;
begin
  if not NSThread.isMainThread then
    exit failedPresentationWithMessage('工作区界面必须在主线程显示');

  if fWindow <> nil then
    exit failedPresentationWithMessage('工作区已经显示');

  var nativeFrame := workspaceFrame as MacWorkspaceFrame;
  if (nativeFrame = nil) or nativeFrame.IsReleased then
    exit failedPresentationWithMessage('无法显示工作区：截图无效');

  if display = nil then
    exit failedPresentationWithMessage('无法显示工作区：显示器信息无效');

  var targetScreen := screenWithIdentifier(display.DisplayId);
  if targetScreen = nil then
    exit failedPresentationWithMessage('无法显示工作区：目标显示器不可用');

  var image := nativeFrame.imageForPresentation;
  if image = nil then
    exit failedPresentationWithMessage('无法显示工作区：截图已经释放');

  var newWindow: MacWorkspaceWindow := nil;
  try
    var screenFrame := targetScreen.frame;
    newWindow := new MacWorkspaceWindow withContentRect(screenFrame) styleMask(NSWindowStyleMask.Borderless) backing(NSBackingStoreType.Buffered) defer(false) screen(targetScreen);
    // The surface keeps the window in a strong ARC field until dismissal has
    // completed. Match the original Cocoa implementation and prevent close()
    // from scheduling a second ownership release for that same window.
    newWindow.releasedWhenClosed := false;
    newWindow.backgroundColor := NSColor.blackColor;
    newWindow.opaque := true;
    newWindow.acceptsMouseMovedEvents := true;
    newWindow.level := NSMainMenuWindowLevel;
    newWindow.collectionBehavior := NSWindowCollectionBehavior.CanJoinAllSpaces or
      NSWindowCollectionBehavior.FullScreenAuxiliary;

    var contentFrame := NSMakeRect(0.0, 0.0, screenFrame.size.width, screenFrame.size.height);
    var contentView := new NSView withFrame(contentFrame);
    var imageView := new MacWorkspaceImageView withFrame(contentFrame);
    imageView.Image := image;
    imageView.wantsLayer := true;
    contentView.addSubview(imageView);

    var hudLabel := NSTextField.labelWithString('100%');
    hudLabel.frame := NSMakeRect(16.0, 16.0, 88.0, 36.0);
    hudLabel.font := NSFont.monospacedDigitSystemFontOfSize(18.0) weight(NSFontWeightSemibold);
    hudLabel.textColor := NSColor.whiteColor;
    hudLabel.alignment := NSTextAlignment.Center;
    hudLabel.wantsLayer := true;
    hudLabel.layer.backgroundColor := NSColor.colorWithWhite(0.0) alpha(0.72).CGColor;
    hudLabel.layer.cornerRadius := 7.0;
    hudLabel.hidden := true;
    contentView.addSubview(hudLabel);

    newWindow.contentView := contentView;
    imageView.CommandRequested := @workspaceCommandRequested;
    imageView.DismissRequested := @workspaceDismissRequested;

    fWindow := newWindow;
    fImageView := imageView;
    fHudLabel := hudLabel;
    fOwnedFrame := nativeFrame;
    fTargetDisplayId := display.DisplayId;
    fDidRequestDisplayDismissal := false;
    beginObservingDisplayForWindow(newWindow) displayId(fTargetDisplayId);

    NSApplication.sharedApplication.activateIgnoringOtherApps(true);
    newWindow.makeKeyAndOrderFront(nil);
    newWindow.makeFirstResponder(imageView);
    result := WorkspacePresentationResult.succeeded;
  except
    on caughtError: Exception do begin
      fIsDismissing := true;
      stopObservingDisplayChanges;
      if newWindow <> nil then
        newWindow.close;

      clearPresentationWithoutReleasingFrame;
      fIsDismissing := false;
      result := failedPresentationWithMessage('无法显示工作区');
    end;
  end;
end;

method MacWorkspaceSurfaceActual.renderTransform(transform: WorkspaceTransform) showHud(showHud: Boolean);
begin
  if not NSThread.isMainThread then begin
    NSOperationQueue.mainQueue.addOperationWithBlock(method
      begin
        renderTransform(transform) showHud(showHud);
      end);
    exit;
  end;

  if (fWindow = nil) or (fImageView = nil) or (transform = nil) then
    exit;

  fImageView.renderTransform(transform);

  if fHudLabel <> nil then begin
    fHudLabel.stringValue := NSString.stringWithFormat('%.0f%%', transform.Scale * 100.0);
    if showHud then
      showHudTemporarily;
  end;
end;

method MacWorkspaceSurfaceActual.dismissPresentation;
begin
  if not NSThread.isMainThread then begin
    NSOperationQueue.mainQueue.addOperationWithBlock(method
      begin
        dismissPresentation;
      end);
    exit;
  end;

  fIsDismissing := true;
  NSLog('Zoomer: beginning workspace window dismissal');
  stopObservingDisplayChanges;
  stopHudFadeTimer;

  if fImageView <> nil then begin
    fImageView.CommandRequested := nil;
    fImageView.DismissRequested := nil;
  end;

  if fWindow <> nil then begin
    NSLog('Zoomer: closing workspace window');
    fWindow.orderOut(nil);
    fWindow.close;
    NSLog('Zoomer: workspace window close returned');
  end;

  releaseOwnedFrame;
  clearPresentationWithoutReleasingFrame;
  fIsDismissing := false;
  NSLog('Zoomer: workspace dismissal completed');
end;

method MacWorkspaceSurfaceActual.workspaceCommandRequested(command: WorkspaceCommand);
begin
  var commandListener := fCommandRequested;
  if assigned(commandListener) then
    commandListener(command);
end;

method MacWorkspaceSurfaceActual.showHudTemporarily;
begin
  stopHudFadeTimer;
  fHudLabel.hidden := false;
  fHudLabel.alphaValue := 1.0;
  fHudFadeTimer := NSTimer.scheduledTimerWithTimeInterval(0.8) repeats(false) &block(method(aTimer: NSTimer)
    begin
      fHudFadeTimer := nil;
      if fHudLabel <> nil then begin
        NSAnimationContext.runAnimationGroup(method(animation: NSAnimationContext)
          begin
            animation.duration := 0.2;
            fHudLabel.animator.alphaValue := 0.0;
          end) completionHandler(nil);
      end;
    end);
end;

method MacWorkspaceSurfaceActual.stopHudFadeTimer;
begin
  if fHudFadeTimer <> nil then begin
    fHudFadeTimer.invalidate;
    fHudFadeTimer := nil;
  end;
end;

method MacWorkspaceSurfaceActual.workspaceDismissRequested;
begin
  requestDismissal;
end;

method MacWorkspaceSurfaceActual.beginObservingDisplayForWindow(window: NSWindow) displayId(targetDisplayId: String);
begin
  var center := NSNotificationCenter.defaultCenter;
  fScreenObserver := center.addObserverForName(NSApplicationDidChangeScreenParametersNotification) object(nil) queue(NSOperationQueue.mainQueue) usingBlock(method(notification: NSNotification)
    begin
      checkTargetDisplayAvailability;
    end);
  fWindowCloseObserver := center.addObserverForName(NSWindowWillCloseNotification) object(window) queue(NSOperationQueue.mainQueue) usingBlock(method(notification: NSNotification)
    begin
      if not fIsDismissing then
        requestDismissal;
    end);
end;

method MacWorkspaceSurfaceActual.stopObservingDisplayChanges;
begin
  var center := NSNotificationCenter.defaultCenter;
  if fScreenObserver <> nil then begin
    center.removeObserver(fScreenObserver);
    fScreenObserver := nil;
  end;

  if fWindowCloseObserver <> nil then begin
    center.removeObserver(fWindowCloseObserver);
    fWindowCloseObserver := nil;
  end;
end;

method MacWorkspaceSurfaceActual.checkTargetDisplayAvailability;
begin
  if (fTargetDisplayId = nil) or fDidRequestDisplayDismissal then
    exit;

  if screenWithIdentifier(fTargetDisplayId) <> nil then
    exit;

  fDidRequestDisplayDismissal := true;
  var disconnectListener := fTargetDisplayDisconnected;
  if assigned(disconnectListener) then
    disconnectListener();
end;

method MacWorkspaceSurfaceActual.screenWithIdentifier(displayId: String): NSScreen;
begin
  if displayId = nil then begin
    result := nil;
    exit;
  end;

  for each screen: NSScreen in NSScreen.screens do begin
    if identifierForScreen(screen) = displayId then
      exit screen;
  end;

  result := nil;
end;

method MacWorkspaceSurfaceActual.identifierForScreen(screen: NSScreen): String;
begin
  var displayNumber := screen.deviceDescription.objectForKey('NSScreenNumber') as NSNumber;
  if displayNumber = nil then begin
    result := nil;
    exit;
  end;

  result := displayNumber.stringValue;
end;

method MacWorkspaceSurfaceActual.failedPresentationWithMessage(message: String): WorkspacePresentationResult;
begin
  result := WorkspacePresentationResult.failedWithFailure(new WorkspaceFailure(WorkspaceFailureCode.PresentationFailed) message(message));
end;

method MacWorkspaceSurfaceActual.releaseOwnedFrame;
begin
  if fOwnedFrame <> nil then begin
    fOwnedFrame.releaseFrame;
    fOwnedFrame := nil;
  end;
end;

method MacWorkspaceSurfaceActual.clearPresentationWithoutReleasingFrame;
begin
  fWindow := nil;
  fImageView := nil;
  fHudLabel := nil;
  fOwnedFrame := nil;
  fTargetDisplayId := nil;
  fDidRequestDisplayDismissal := false;
end;

method MacWorkspaceSurfaceActual.requestDismissal;
begin
  var dismissListener := fDismissRequested;
  if assigned(dismissListener) then
    dismissListener();
end;

end.
